sdc/native/stable_sort.cpp

```diff
@@ -0,0 +1,333 @@
+#include "utils.hpp"
+#include "tbb/parallel_invoke.h"
+#include <iostream>
+
+using namespace utils;
+
+namespace
+{
+
+template<class T>
+struct buffer_queue
+{
+    using v_type = T;
+    v_type* head;
+    v_type* tail;
+
+    buffer_queue(v_type* _head, int size, int _item_size)
+    {
+        (void)_item_size;
+
+        head = _head;
+        tail = head + size;
+    }
+
+    inline v_type* pop() { return head++; }
+
+    inline bool not_empty() { return head < tail; }
+
+    inline void push(v_type* val) { *(tail++) = *val; }
+
+    inline int size() { return tail - head; }
+
+    inline int copy_size() { return size(); }
+};
+
+template<>
+struct buffer_queue<void>
+{
+    using v_type = void;
+    quant* head;
+    quant* tail;
+    const int item_size;
+
+    buffer_queue(v_type* _head, int size, int _item_size) : item_size(_item_size)
+    {
+        head = as_quant(_head);
+        tail = head + size*item_size;
+    }
+
+    inline v_type* pop()
+    {
+        auto _h = head;
+        head += item_size;
+        return _h;
+    }
+
+    inline bool not_empty() { return head < tail; }
+
+    inline void push(v_type* val)
+    {
+        std::copy_n(as_quant(val),
+                    item_size,
+                    tail);
+
+        tail += item_size;
+    }
+
+    inline int size() { return copy_size()/item_size; }
+
+    inline int copy_size() { return tail - head; }
+
+    inline quant* as_quant(v_type* data) { return reinterpret_cast<quant*>(data); }
+};
+
+template<class T>
+inline void merge_sorted_main_loop(buffer_queue<T>& left, buffer_queue<T>& right, buffer_queue<T>& out, void* compare = nullptr)
+{
+    if (compare)
```

Comment:
You can write this shorter, like
```auto less = compare ? reinterpret_cast<compare_func>(compare) : [](T* r, T* l){ return *r < *l; }```

Reply:
calling `comapre` instead of using simple `<` causes 20-30% of performance degradation